modules/events.js

```diff
@@ -32,286 +32,403 @@
+ * No checks are made to see if the listener has already been added.
+ * Multiple calls passing the same combination of event and listener
+ * will result in the listener being added, and called, multiple times.
+ * EventEmitter is defined in src/node_events.cc
+ * EventEmitter.prototype.emit() is also defined there.
+ *
+ * @param  String    type      An event name.
+ * @param  function  filterFn  An options callback to apply on event.
+ * @return EventEmitter        A reference to the EventEmitter so calls can be chained
+ * @throws {CasperError}       If invokation failed.
+ */
 EventEmitter.prototype.addListener = function addListener(type, listener) {
-  if ('function' !== typeof listener) {
-    throw new CasperError('addListener only takes instances of Function');
-  }
-
-  if (!this._events) this._events = {};
-
-  // To avoid recursion in the case that type == "newListeners"! Before
-  // adding it to the listeners, first emit "newListeners".
-  this.emit('newListener', type, listener);
-
-  if (!this._events[type]) {
-    // Optimize the case of one listener. Don't need the extra array object.
-    this._events[type] = listener;
-  } else if (isArray(this._events[type])) {
-
-    // If we've already got an array, just append.
-    this._events[type]['fail' === type ? 'unshift' : 'push'](listener);
-
-    // Check for listener leak
-    if (!this._events[type].warned) {
-      var m;
-      if (this._maxListeners !== undefined) {
-        m = this._maxListeners;
-      } else {
-        m = defaultMaxListeners;
-      }
-
-      if (m && m > 0 && this._events[type].length > m) {
-        this._events[type].warned = true;
-        console.error('(node) warning: possible EventEmitter memory ' +
-                      'leak detected. %d listeners added. ' +
-                      'Use emitter.setMaxListeners() to increase limit.',
-                      this._events[type].length);
-        console.trace();
-      }
+    "use strict";
+    if (typeof listener !== 'function') {
+        throw new CasperError('addListener only takes instances of Function');
     }
-  } else {
-    // Adding the second element, need to change to array.
-    this._events[type] = 'fail' === type ? [listener, this._events[type]] : [this._events[type], listener];
-  }
 
-  return this;
+    if (!this._events) {
+        this._events = {};
+    }
+
+    // To avoid recursion in the case that type == "newListeners"! Before
+    // adding it to the listeners, first emit "newListeners".
+    this.emit('newListener', type, listener);
+
+    if (!this._events[type]) {
+        // Optimize the case of one listener. Don't need the extra array object.
+        this._events[type] = listener;
+    } else if (isArray(this._events[type])) {
+        // If we've already got an array, just append.
+        this._events[type][ type === 'fail' ? 'unshift' : 'push'](listener);
+
+        // Check for listener leak
+        if (!this._events[type].warned) {
+            var m;
```

Comment:
can you use a more explicit variable name ? Are we ok this is a kind a stack of listeners atm ?